source/GameInterface/Utils/AutoSync/Dynamic/DataClassGenerator.cs

```diff
@@ -0,0 +1,151 @@
+﻿using GameInterface.Utils.AutoSync.Template;
+using HarmonyLib;
+using ProtoBuf;
+using System;
+using System.Collections.Generic;
+using System.IO.Pipes;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace GameInterface.Utils.AutoSync.Dynamic;
+
+/// <summary>
+/// Constructs a new data class for a given type.
+/// </summary>
+/// <remarks>
+/// Data class constists of two fields: StringId and Value.
+/// </remarks>
+public class DataClassGenerator
+{
+    private readonly ModuleBuilder moduleBuilder;
+    private int protoMemberCounter = 1;
+
+    public DataClassGenerator(ModuleBuilder moduleBuilder)
+    {
+        this.moduleBuilder = moduleBuilder;
+    }
+
+    public Type GenerateClass(Type dataType, string name)
+    {
+        // Creates type builder for the new data class
+        TypeBuilder typeBuilder = moduleBuilder.DefineType(
+            $"{name}Data",
+            TypeAttributes.Public | TypeAttributes.Sealed | TypeAttributes.AnsiClass | TypeAttributes.AutoLayout | TypeAttributes.BeforeFieldInit,
+            typeof(object)
```

Comment:
Rather then being a typeof(object). For tracking wouldn't something like this be useful? so you can verify the field. unless its just string, int, etc may be pointless.

```c#
public class CustomDataTypeGenerator : Object { }

...

        TypeBuilder typeBuilder = moduleBuilder.DefineType(
            $"{name}Data",
            TypeAttributes.Public | TypeAttributes.Sealed | TypeAttributes.AnsiClass | TypeAttributes.AutoLayout | TypeAttributes.BeforeFieldInit,
            typeof(CustomDataTypeGenerator)
        );
```

Reply:
This actually inherits from an interface `IAutoSyncData`.
typeof(object) can be removed